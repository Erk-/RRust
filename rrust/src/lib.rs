#[macro_export]
macro_rules! rfn {
    ($name:ident, ($($param:ident: $party:ty),* $(,)?), $code:block) => {
        struct $name;

        impl $name {
            fn forward($($param:$party),*) {
                ::rrust::forward! {
                    $code
                };
            }
            fn backwards($($param:$party),*) {
                ::rrust::reverse! {
                    $code
                };
            }
        }
    }
}

#[macro_export]
macro_rules! rif {
    ($before:expr, $then:block, $else:block, $after:expr) => {
        if $before {
            ::rrust::forward! {
                $then
            };
            assert!($after);
        } else {
            ::rrust::forward! {
                $else
            };
            assert!(!($after));
        }
    };
    ($before:expr, $then:block, $after:expr) => {
        if $before {
            ::rrust::forward! {
                $then
            };
            assert!($after);
        }
    };
}

#[macro_export]
macro_rules! _reverse_rif {
    ($before:expr, $then:block, $else:block, $after:expr) => {
        if $after {
            ::rrust::reverse! {
                $then
            };
            assert!($before);
        } else {
            ::rrust::reverse! {
                $else
            };
            assert!(!($before));
        }
    };
    ($before:expr, $then:block, $after:expr) => {
        if $after {
            ::rrust::reverse! {
                $then
            };
            assert!($before);
        }
    };
}

#[macro_export]
macro_rules! rloop {
    ($from:expr, $do:block, $loop:block, $until:expr) => {
        assert!($from);
        ::rrust::forward! {
            $do
        };
        while !$until {
            ::rrust::forward! {
                $loop
            };
            assert!(!$from);
            ::rrust::forward! {
                $do
            };
        }
    };
    ($from:expr, $loop:block, $until:expr) => {
        assert!($from);
        while !$until {
            ::rrust::forward! {
                $loop
            };
            assert!(!$from);
        }
    };
}

#[macro_export]
macro_rules! _reverse_rloop {
    ($from:expr, $do:block, $loop:block, $until:expr) => {
        assert!($until);
        ::rrust::reverse! {
            $do;
        };
        while !$from {
            ::rrust::reverse! {
                $loop;
            };
            assert!(!$until);
            ::rrust::reverse! {
                $do;
            };
        }
    };
    ($from:expr, $loop:block, $until:expr) => {
        assert!($until);
        while !$from {
            ::rrust::reverse! {
                $loop
            };
            assert!(!$until);
        }
    };
}

#[doc(hidden)]
pub use rrust_macro::{forward, reverse};

#[macro_export]
macro_rules! delocal {
    ($name:ident, $e:expr) => {
        if $name != $e {
            panic!("Delocal failed {} != {}", $name, $e);
        }
        drop($name);
    };
}

