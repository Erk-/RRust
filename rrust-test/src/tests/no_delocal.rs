use rrust::rfn;

rfn!(NoDelocal, (x1: &mut i32), {
    let i = 1
    *x1 += i;
});

fn main() {
    let mut x = 0;

    NoDelocal::forward(&mut x);
    NoDelocal::backwards(&mut x);
}
