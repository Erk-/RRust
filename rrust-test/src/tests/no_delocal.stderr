error: proc macro panicked
 --> src/tests/no_delocal.rs:3:1
  |
3 | / rfn!(NoDelocal, (x1: &mut i32), {
4 | |     let i = 1
5 | |     *x1 += i;
6 | | });
  | |__^
  |
  = help: message: The following local(s) need to be consumed by delocal!: ["i"]
  = note: this error originates in the macro `rfn` (in Nightly builds, run with -Z macro-backtrace for more info)

error: proc macro panicked
 --> src/tests/no_delocal.rs:3:1
  |
3 | / rfn!(NoDelocal, (x1: &mut i32), {
4 | |     let i = 1
5 | |     *x1 += i;
6 | | });
  | |__^
  |
  = help: message: The following local(s) need to be consumed by delocal!: ["i"]
  = note: this error originates in the macro `rfn` (in Nightly builds, run with -Z macro-backtrace for more info)
